File: src/OnePhasedFamily.cpp
#include <algorithm>
#include <cassert>
#include "../include/OnePhasedFamily.h"
#include "../include/VCFImpHeteroHomo.h"
#include "../include/VCFOneParentImputed.h"
#include "../include/VCFOneParentImputedRough.h"
#include "../include/VCFHeteroImpHomo.h"
#include "../include/VCFSmallFillable.h"
#include "../include/Pedigree.h"
#include "../include/KnownFamily.h"
#include "../include/Genotype.h"
#include "../include/common.h"

using namespace std;


//////////////////// OnePhasedFamily ////////////////////

int OnePhasedFamily::get_gt_type(const VCFRecord *record, int i) {
	if(record->is_hetero(i))
		return 0;	// 0/1
	else if(record->get_gt(i).c_str()[0] == '0')
		return -1;	// 0/0
	else
		return 1;	// 1/1
}

std::pair<ParentComb, FillType> OnePhasedFamily::classify_record(
													const VCFRecord *record) {
	if(record->is_NA(0) || record->is_NA(1))
		return make_pair(ParentComb::PNA, FillType::IMPUTABLE);
	
	const int	mat_gt_type = get_gt_type(record, 0);
	const int	pat_gt_type = get_gt_type(record, 1);
	if(mat_gt_type == 0) {
		switch(pat_gt_type) {
			case 0:  return make_pair(ParentComb::P01x01, FillType::IMPUTABLE);
			case -1: return make_pair(ParentComb::P00x01, FillType::MAT);
			default: return make_pair(ParentComb::P01x11, FillType::MAT);
		}
	}
	else if(mat_gt_type == -1) {
		switch(pat_gt_type) {
			case 0:  return make_pair(ParentComb::P00x01, FillType::PAT);
			case -1: return make_pair(ParentComb::P00x00, FillType::FILLED);
			default: return make_pair(ParentComb::P00x11, FillType::FILLED);
		}
	}
	else {
		switch(pat_gt_type) {
			case 0:  return make_pair(ParentComb::P01x11, FillType::PAT);
			case -1: return make_pair(ParentComb::P00x11, FillType::FILLED);
			default: return make_pair(ParentComb::P11x11, FillType::FILLED);
		}
	}
}

array<vector<VCFFillableRecord *>, 4>
OnePhasedFamily::classify_records(const vector<VCFFamilyRecord *>& records) {
	// hetero x hetero, homo x hetero, hetero x homo, homo x homo
	array<vector<VCFFillableRecord *>, 4>	rss;
	for(size_t index = 0; index < records.size(); ++index) {
		VCFFamilyRecord	*record = records[index];
		const auto	p = classify_record(record);
		const ParentComb	comb = p.first;
		const FillType		type = p.second;
		auto	r = new VCFFillableRecord(record->get_v(),
											record->get_samples(),
											index, type, comb);
		assert(static_cast<int>(type) < 4);
		rss[static_cast<int>(type)].push_back(r);
	}
	return rss;
}

VCFHeteroHomoOnePhased *OnePhasedFamily::create(const vector<STRVEC>& header,
										const STRVEC& samples,
										const vector<VCFFillableRecord *>& rs,
										bool is_mat_hetero, bool is_mat_imputed,
										const Map& gmap) {
	if(is_mat_hetero == is_mat_imputed)
		return new VCFImpHeteroHomo(header, samples, rs, is_mat_hetero, gmap);
	else
		return new VCFHeteroImpHomo(header, samples, rs, is_mat_hetero, gmap);
}

bool OnePhasedFamily::compare_record(const VCFRecord *a, const VCFRecord *b) {
	return a->pos() < b->pos();
}

VCFSmallFillable *OnePhasedFamily::merge_vcf(
					const std::vector<STRVEC>& header, const STRVEC& samples,
					const array<vector<VCFFillableRecord *>, 4>& rss) {
	vector<VCFFillableRecord *>	rs;
	for(int i = 0; i < 4; ++i) {
		for(auto p = rss[i].begin(); p != rss[i].end(); ++p)
			rs.push_back(*p);
	}
	std::sort(rs.begin(), rs.end(), OnePhasedFamily::compare_record);
	return new VCFSmallFillable(header, samples, rs);
}

VCFSmallBase *OnePhasedFamily::impute(const Family& family, VCFFamily *vcf,
											const STRVEC& non_imputed_parents,
											const Map& gmap) {
	const auto&	header = vcf->get_header();
	const STRVEC&	samples = family.get_samples();
	const auto	rss = classify_records(vcf->get_family_records());
	const bool	is_mat_imp = std::find(non_imputed_parents.begin(),
										non_imputed_parents.end(),
										family.get_pat())
												!= non_imputed_parents.end();
	const auto&	rs_mat = rss[static_cast<int>(FillType::MAT)];
	auto	*mat_vcf = create(header, samples, rs_mat, true, is_mat_imp, gmap);
	mat_vcf->impute();
	delete mat_vcf;
	const auto&	rs_pat = rss[static_cast<int>(FillType::PAT)];
	auto	*pat_vcf = create(header, samples, rs_pat, false, is_mat_imp, gmap);
	pat_vcf->impute();
	delete pat_vcf;
	auto	*merged_vcf = merge_vcf(header, samples, rss);
	merged_vcf->modify(1);
	return merged_vcf;
}

vector<vector<int>> OnePhasedFamily::filter_ref_haps(
							const vector<vector<int>>& ref_haps,
							const Family *family, bool is_mat_imputed,
							const VCFFamily *vcf) {
	const size_t	c = is_mat_imputed ? 10 : 9;
	// collect homo indices
	vector<size_t>	ks;
	for(size_t k = 0; k < vcf->size(); ++k) {
		if(Genotype::is_homo(vcf->get_record(k)->get_gt(c-9)))
			ks.push_back(k);
	}
	
	// tiling
	const size_t	L = 10;
	vector<size_t>	firsts;		// ranges [first, first+L)
	for(size_t first = 0; first < ks.size() - L; first += L) {
		firsts.push_back(first);
	}
	if(ks.size() % L != 0 && ks.size() > L)
		firsts.push_back(ks.size() - L);
	
	set<size_t>	ref_indices_set;
	for(auto p = firsts.begin(); p != firsts.end(); ++p) {
		int	code = 0;
		for(size_t j = *p; j < *p + L; ++j) {
			const string&	gt = vcf->get_record(ks[j])->get_gt(c-9);
			code = code * 2 + (gt.c_str()[0] - '0');
		}
		
		for(size_t i = 0; i < ref_haps.size(); ++i) {
			int	ref_code = 0;
			for(size_t j = *p; j < *p + L; ++j) {
				code = code * 2 + ref_haps[i][ks[j]];
			}
			if(ref_code == code)
				ref_indices_set.insert(i);
		}
	}
	
	vector<vector<int>>	new_ref_haps;
	for(auto p = ref_indices_set.begin(); p != ref_indices_set.end(); ++p) {
		new_ref_haps.push_back(ref_haps[*p]);
	}
	return new_ref_haps;
}

// Is the computational cost sufficiently small even when using ref in HMM?
bool OnePhasedFamily::is_small(const Family *family,
								const vector<vector<int>>& ref_haps) {
	const size_t	N = family->num_progenies();
	if(N > 1)
		return false;
	
	const size_t	M = ref_haps[0].size();
	const size_t	NH = ref_haps.size();
	const size_t	R = NH * NH * (2*NH + 2*N - 1) << (N*2);
	return R * M < 100000000 && R < 100000;		// 10^8 & 10^5
}

bool OnePhasedFamily::is_small_ref(const vector<vector<int>>& ref_haps) {
	const size_t	M = ref_haps[0].size();
	const size_t	NH = ref_haps.size();
	const size_t	R = NH * NH * (2*NH - 1);
	return R * M < 100000000 && R < 100000;		// 10^8 & 10^5
}

void OnePhasedFamily::impute_small_in_thread(void *config) {
	auto	*c = (ConfigThread *)config;
	const auto&	vcfs = c->vcfs;
	const size_t	n = vcfs.size();
	for(size_t i = c->first; i < n; i += c->num_threads) {
		vcfs[i]->impute();
	}
}

void OnePhasedFamily::impute_small_VCFs(
						vector<VCFOneParentImputed *>& vcfs, int T) {
	// VCFOneParentImputed is heavy for imputation,
	// so make it multi-threaded and impute in order of processing load.
	std::sort(vcfs.begin(), vcfs.end(),
				[](const VCFOneParentImputed * a,
					const VCFOneParentImputed * b) {
						return a->num_progenies() > b->num_progenies();
	});
	
	vector<ConfigThread *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThread(i, T, vcfs);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
			(void *(*)(void *))&impute_small_in_thread, (void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		impute_small_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
}

VCFSmallBase *OnePhasedFamily::impute_by_parent(
									const VCFSmall *orig_vcf,
									const VCFSmall *imputed_vcf,
									const vector<vector<int>>& ref_haps,
									const vector<const KnownFamily *>& families,
									const STRVEC& non_imputed_parents,
									const Map& gmap, int num_threads) {
	const size_t	N = families.size();
	vector<const VCFSmallBase *>	vcfs(N);
	vector<VCFOneParentImputed *>	small_vcfs;
	for(size_t i = 0; i < N; ++i) {
		const KnownFamily	*family = families[i];
		auto	*vcf1 = VCFFamily::create_by_two_vcfs(imputed_vcf,
											orig_vcf, family->get_samples());
		const bool	is_mat_imputed = std::find(non_imputed_parents.begin(),
											   non_imputed_parents.end(),
											   family->get_pat())
										!= non_imputed_parents.end();
		if(is_small(family, ref_haps)) {
			auto	*vcf = new VCFOneParentImputed(vcf1->get_header(),
												   family->get_samples(),
												   vcf1->get_family_records(),
												   ref_haps,
												   is_mat_imputed, gmap, 0.01);
			small_vcfs.push_back(vcf);
			vcfs[i] = vcf;
			// The records are being reused,
			// so the original VCF is emptied before being deleted.
			vcf1->clear_records();
		}
		else if(is_small_ref(ref_haps)) {
			auto	*vcf2 = new VCFOneParentImputedRough(vcf1->get_header(),
												family->get_samples(),
												vcf1->get_family_records(),
												ref_haps,
												is_mat_imputed, gmap, 0.01);
			vcf2->impute();
			vcfs[i] = vcf2;
			vcf1->clear_records();
		}
		else {
			auto	*imputed_vcf = impute(*family, vcf1,
											non_imputed_parents, gmap);
			vcfs[i] = imputed_vcf;
		}
		delete vcf1;
	}
	
	// Small VCFs are heavy to process, so it will be parallelized.
	impute_small_VCFs(small_vcfs, num_threads);
	auto	*new_vcf = VCFSmall::join(vcfs, orig_vcf->get_samples());
	Common::delete_all(vcfs);
	return new_vcf;
}

File: python/OnePhasedFamily.py
from __future__ import annotations

# coding: utf-8
# OnePhasedFamily.py
# 片親がphasingされて片親は分っているがphasingされていない家系を補完する

from functools import reduce
from collections import defaultdict, Counter
from typing import List, Tuple, Optional, IO, Dict, Iterator, Sequence

from VCFFamily import *
from VCFSmallFillable import *
from group import Groups
from RecordSet import RecordSet
from VCFImpFamily import FillType
from VCFHeteroHomoPP import *
from VCFImpHeteroHomo import *
from VCFHeteroImpHomo import *
from Map import *
import ClassifyRecord
from TypeDeterminer import *
from VCFOneParentImputed import VCFOneParentImputed
from VCFOneParentImputedRough import VCFOneParentImputedRough
from Genotype import Genotype


# VCFHeteroHomoPPを使わずにこれを使う あとで統合する
def classify_record(record: VCFRecord) -> tuple[ParentComb, FillType]:
	if record.is_NA(0) or record.is_NA(1):
		return (ParentComb.PNA, FillType.IMPUTABLE)
	
	i = 0 if record.v[9][0] != record.v[9][2] else 1
	j = 0 if record.v[10][0] != record.v[10][2] else 1
	if (i, j) == (0, 0):		# 0/1 x 0/1
		return (ParentComb.P01x01, FillType.IMPUTABLE)
	elif (i, j) == (1, 0):		# 0/0 x 0/1 or 1/1 x 0/1
		if record.v[9][0] == '0':
			return (ParentComb.P00x01, FillType.PAT)
		else:
			return (ParentComb.P01x11, FillType.PAT)
	elif (i, j) == (0, 1):		# 0/1 x 0/0 or 0/1 x 1/1
		if record.v[10][0] == '0':
			return (ParentComb.P00x01, FillType.MAT)
		else:
			return (ParentComb.P01x11, FillType.MAT)
	else:
		if record.v[9][0] == '0' and record.v[10][0] == '0':	# 0/0 x 0/0
			return (ParentComb.P00x00, FillType.FILLED)
		elif record.v[9][0] == '1' and record.v[10][0] == '1':	# 1/1 x 1/1
			return (ParentComb.P11x11, FillType.FILLED)
		else:													# 0/0 x 1/1
			return (ParentComb.P00x11, FillType.FILLED)

def classify_records(records: Sequence[VCFFamilyRecord]
								) -> list[list[VCFFillableRecord]]:
	# ヘテロ×ヘテロ, ホモ×ヘテロ, ヘテロ×ホモ, ホモ×ホモ
	rss: list[list[VCFFillableRecord]] = [ [] for _ in range(4) ]
	for index, record in enumerate(records):
		pair, type = classify_record(record)
		rss[type.value].append(VCFFillableRecord(record.v, record.samples,
														index, type, pair))
	return rss

def create(header: list[list[str]], rs: list[VCFFillableRecord],
						is_mat_hetero: bool, is_mat_imputed: bool,
						gmap: Map) -> VCFHeteroHomoOnePhased:
	if is_mat_hetero == is_mat_imputed:
		return VCFImpHeteroHomo(header, rs, is_mat_hetero, gmap)
	else:
		return VCFHeteroImpHomo(header, rs, is_mat_hetero, gmap)

def merge_vcf(rss: list[list[VCFFillableRecord]],
							header: list[list[str]]) -> VCFSmallFillable:
	rs = rss[0] + rss[1] + rss[2] + rss[3]
	rs.sort(key=lambda r: r.pos())
	return VCFSmallFillable(header, rs)

# HMMにrefを使っても計算量が十分に小さいか
def is_small(family: Family, ref_haps: list[list[int]]) -> bool:
	N = family.num_progenies()
	M = len(ref_haps[0])				# マーカー数
	NH = len(ref_haps)
	R = NH**2 * 4**N * (2*NH + 2*N - 1)
	return R * M < 10**8 and R < 10**5

def is_small_ref(ref_haps: list[list[int]]) -> bool:
	M = len(ref_haps[0])				# マーカー数
	NH = len(ref_haps)
	R = NH**2 * (2*NH - 1)
	return R * M < 10**8 and R < 10**5

def impute(family: Family, vcf: VCFFamily,
					unimputed_parents: list[str], gmap: Map) -> VCFSmallBase:
	header = vcf.header
	rss = classify_records(vcf.records)
	is_mat_imp = family.pat in unimputed_parents
	mat_vcf = create(header, rss[FillType.MAT.value], True, is_mat_imp, gmap)
	mat_vcf.impute()
	pat_vcf = create(header, rss[FillType.PAT.value], False, is_mat_imp, gmap)
	pat_vcf.impute()
	merged_vcf = merge_vcf(rss, header)
	merged_vcf.modify(False)
	return merged_vcf

def impute_by_parent(orig_vcf: VCFSmall, imputed_vcf: VCFSmall,
										ref_haps: list[list[int]],
										families: list[Family],
										non_imputed_parents: list[str],
										gmap: Map) -> VCFSmallBase:
	vcfs: list[VCFSmallBase] = []
	for family in families:
		vcf1 = VCFFamily.create_by_two_vcfs(imputed_vcf,
											orig_vcf, family.samples())
		is_mat_imputed = family.pat in non_imputed_parents
		if is_small(family, ref_haps):
			vcf = VCFOneParentImputed(vcf1.header, vcf1.records,
										ref_haps, is_mat_imputed, gmap)
			vcf.impute()
			vcfs.append(vcf)
		elif is_small_ref(ref_haps):
			vcf2 = VCFOneParentImputedRough(vcf1.header, vcf1.records,
										ref_haps, is_mat_imputed, gmap)
			vcf2.impute()
			vcfs.append(vcf2)
		else:
			imputed_vcf1 = impute(family, vcf1, non_imputed_parents, gmap)
			vcfs.append(imputed_vcf1)
	
	new_vcf = VCFSmall.join(vcfs, orig_vcf.samples)
	return new_vcf

__all__ = ['impute_by_parent']
